atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1114,24 +1114,27 @@ private int estimateSize(List<RowResult<Value>> rangeRows) {
                 } else {
                     // The value has a commit timestamp less than our start timestamp, and is visible and valid.
                     if (value.getContents().length != 0) {
-                        found++;
                         results.put(key, transformer.apply(value));
+                        keysAddedBuilder.add(key);
                     }
                 }
             }
         }
-        count.set(found);
+        Set<Cell> keysAddedToResults = keysAddedBuilder.build();
+        count.addAndGet(keysAddedToResults.size());
 
         if (!keysToDelete.isEmpty()) {
             // if we can't roll back the failed transactions, we should just try again
             if (!rollbackFailedTransactions(tableRef, keysToDelete, commitTimestamps, defaultTransactionService)) {
+                rawResults.keySet().removeAll(keysAddedToResults);
                 return rawResults;
             }
         }
 
         if (!keysToReload.isEmpty()) {
             Map<Cell, Value> nextRawResults = keyValueService.get(tableRef, keysToReload);
             validateExternalAndCommitLocksIfNecessary(tableRef, getStartTimestamp());
+            nextRawResults.keySet().removeAll(keysAddedToResults);
```

Comment:
Again, gotta be careful here - `kvs.get(...)` might return an immutable map (it does on Cassandra KVS!)